### DISCUSSION TITLE
Why do abseil's Mutex sleep?

### DISCUSSION THREAD

**Discussion by deadalnix** [abseil/abseil-cpp#1374] [General]
I was checking abseil's mutex implementation. When it takes too long to acquire a mutex, abseil [calls sleep](https://github.com/abseil/abseil-cpp/blob/master/absl/synchronization/mutex.cc#L201) . Because the thread has no idea how long it'll have to wait, it has to wake up regularly, and is going to oversleep by roughly half the time it sleeps on average.

Would it be preferable to use system primitives such as futexes? This is perplexing to me, because the spinlock uses futexes so it's not like it is not available or something.

Does any one know the reason why the lock/mutex are designed in this way?

Thanks in advance.

**Comment by derekmauro** [comment]
In some cases `absl::Mutex` spins or sleeps, and in other cases it does use system primitives like futexes. If you check the code in `mutex.cc`, `PerThreadSem::Wait()` eventually uses a futex on Linux.

The logic to decide when to spin and when to use the futex is quite complicated. The main idea is to use the futex as the slow blocking path, but also to try to avoid entering the kernel when a wakeup is unnecessary.

**Comment by deadalnix** [comment]
You are right, I missed this. It's buried quite a few level deep and I missed it.

**Comment by abcd1424** [comment]
Pls  unsubcribed me
